Add description of role of infrastructure in package building (#2150)
conda-forge/conda-forge.github.io: docs/maintainer/infrastructure.md

- `conda-forge-ci-setup-feedstock` provides the code needed to prepare and homogenize the CI runners across providers.
  It also does some checks before the artifacts are uploaded to `cf-staging`.
- `conda-forge-pinning-feedstock` defines which versions are supported for a number of runtimes and libraries, as well as the compilers used for certain languages and platforms.
- `docker-images` builds the standardized container images for Linux runners.
  This repository has additional authentication needs for Docker Hub, Quay.io.

The pipelines can run on several CI providers supported by `conda-smithy`, including:

- Azure DevOps Pipelines
- Travis CI
- Circle CI
- Appveyor
- Self-hosted Github Actions runners

Registration of hooks and triggers is also done by the `conda-smithy` app.

:::tip
`conda-smithy` supports more CI providers.
Check [its repository](https://github.com/conda-forge/conda-smithy) for more details.
:::

Authenticated services involved:

- Anaconda.org uploads to `cf-staging`

### Package validation and publication

Once built on `main` (or other branches), the conda packages are uploaded to an intermediary channel named `cf-staging`.

> From there, the packages are downloaded by the validation server and, if successful, copied over to `conda-forge` itself.
> 
> - The validation logic is defined at `conda-forge/artifact-validation`
> - If problematic, the results of the validation are posted as issues in the same repo.
> - This logic runs at `conda-forge/conda-forge-webservices`.
>   This web app also copies the artifacts from `cf-staging` to `conda-forge`.
> - Part of the validation includes checking for cross-package clobbering.
>   The list of authorized feedstocks per package name is maintained at `conda-forge/feedstock-outputs`.
> - Some further analysis might be performed _after_ publication.
> 
> Authenticated services involved:
> 
> - Anaconda.org uploads to `conda-forge`
> - The `conda-forge-webservices` app deployment itself (currently at Heroku)
> - (?) Post new issues to `conda-forge/artifact-validation`

Reviewer: ```suggestion
From there, our webservices (`conda-forge/conda-forge-webservices`) does the following:

- The logic checks the feedstock token to authenticate a legitimate request.
- The logic checks that the hash sum of the package on `cf-staging` against 
  the value computed in the CI to ensure the artifact to be copied is the same.
- The logic checks that the feedstock is allowed to push the package using 
  the `conda-forge/feedstock-outputs` repo.
- If all three checks pass, the webservices copies the artifacts from `cf-staging` to `conda-forge`.

Authenticated services involved:

- Anaconda.org uploads to `conda-forge` and `cf-staging`
- The `conda-forge-webservices` app deployment itself (currently at Heroku)
```

Once this is done correctly and the package is uploaded,
you can then request the conda-forge core devs to archive the old feedstock.

> ## How package building interacts with infrastructure

Reviewer 2: This is also explaining the "life cycle" of the package. Maybe a different title is useful. Some suggestions:

- Life cycle of a package and the involved infrastructure
- The different stages of package building and the infrastructure used

> Packages in conda-forge are always built through CI.

Reviewer 2: ```suggestion
Packages in conda-forge are always built through CI.
```

This deserves a footnote or info admonition mentioning [CFEP-03](https://github.com/conda-forge/cfep/blob/main/cfep-03.md).

Author: Good call. I added a footnote.

However, when a new package enters conda-forge for the first time, it does so via a pull request in the [`staged-recipes` repository](#staged-recipes), whereas every new build of the package after that is built in its repository, the so-called feedstock.
Both places use slightly different CI setups and interact with the infrastructure accordingly.
Hence, we first describe the interaction at the start of a new package and then for existing packages in their respective feedstocks.

### Initial submission to staged-recipes

The `conda-forge/staged-recipes` repository uses several pieces of infrastructure.

On pull requests:

- Package building pipelines. These are slightly different than the ones running in feedstocks (they are not automatically generated by `conda-smithy`, but they do use the same underlying components).
- The linter is provided by `conda-smithy recipe-lint`, run by `@conda-forge-linter`.
- Auto-labeling logic, run by Github Actions workflows.

On pushes to `main`:

> - Create and initialize the new feedstocks.

Reviewer 2: This is delegated to an external repo: [admin-requests](https://github.com/conda-forge/admin-requests/blob/main/.github/workflows/create_feedstocks.yml).

Authenticated services involved:

> - Github, with permissions for:
>   - Repository creation
> - Azure Pipelines
> - Travis CI
> - Other CI providers?

Reviewer 2: ```suggestion
- Github, with permissions for:
  - PR labeling
- Azure Pipelines
```

Check .github/workflows for details.

### Feedstock changes

A feedstock can receive changes for several reasons.

Pushes to `main` or other branches:

- The automated initialization commits following approval in `staged-recipes`.

>   These are generated by `conda-smithy`.

Reviewer 2: ```suggestion
  These are generated by `conda-smithy` and pushed by the automation in `admin-requests`.
```

- Automated maintenance commits triggered from `admin-migrations`.

> - Rerender requests are handled by instances of `conda-forge/webservices-dispatch-action` and triggered by the admin web service.

Reviewer 2: The admin web service is... the heroku app?

The pipelines that build conda packages are used for both pull requests and push events in `main` and other branches.
The only difference is that the packages built during a pull request are not uploaded to the staging channel.
Maintaining these up-to-date across all feedstocks involves several repositories:

- `conda-smithy` is in charge of generating the CI pipelines themselves, together with the supporting scripts and configuration files.
  These pipelines and scripts can rely on code and data defined in the repositories below.
- `conda-forge-ci-setup-feedstock` provides the code needed to prepare and homogenize the CI runners across providers.
  It also does some checks before the artifacts are uploaded to `cf-staging`.
- `conda-forge-pinning-feedstock` defines which versions are supported for a number of runtimes and libraries, as well as the compilers used for certain languages and platforms.
- `docker-images` builds the standardized container images for Linux runners.
  This repository has additional authentication needs for Docker Hub, Quay.io.

The pipelines can run on several CI providers supported by `conda-smithy`, including:

- Azure DevOps Pipelines
- Travis CI

> - Circle CI
> - Appveyor

Reviewer 2: ```suggestion
- Self-hosted Github Actions runners
```

Registration of hooks and triggers is also done by the `conda-smithy` app.

:::tip
`conda-smithy` supports more CI providers.
Check [its repository](https://github.com/conda-forge/conda-smithy) for more details.
:::

Authenticated services involved:

- Anaconda.org uploads to `cf-staging`

> ### Package validation and publication

Reviewer 2: This might not be up-to-date. I think @beckermr adjusted this part of the infra not long ago.

Author: @beckermr, could you double-check whether this section needs adaptation?

Reviewer: done!

Once built on `main` (or other branches), the conda packages are uploaded to an intermediary channel named `cf-staging`.
From there, the packages are downloaded by the validation server and, if successful, copied over to `conda-forge` itself.

- The validation logic is defined at `conda-forge/artifact-validation`
- If problematic, the results of the validation are posted as issues in the same repo.
- This logic runs at `conda-forge/conda-forge-webservices`.
  This web app also copies the artifacts from `cf-staging` to `conda-forge`.
- Part of the validation includes checking for cross-package clobbering.
  The list of authorized feedstocks per package name is maintained at `conda-forge/feedstock-outputs`.
- Some further analysis might be performed _after_ publication.

Authenticated services involved:

- Anaconda.org uploads to `conda-forge`
- The `conda-forge-webservices` app deployment itself (currently at Heroku)
- (?) Post new issues to `conda-forge/artifact-validation`

### Post-publication

Once uploaded to anaconda.org/conda-forge, packages are not immediately available to CLI clients.
They have to be replicated in the Content Distribution Network (CDN).

> This step usually takes less than 15 minutes.

Reviewer 2: ```suggestion
This step usually should ideally take around 15 minutes. In some circumstances, longer delays are possible. Check [conda-forge.org/status](https://conda-forge.org/status) in case of doubt.
```

After CDN replication, most packages available on anaconda.org/conda-forge won't suffer any further modifications.
However, in some cases, maintainers might need to perform some actions on the published packages:

- Patching their repodata
- Marking them as broken

#### Repodata patch

The metadata for `conda` packages is initially contained in each package archive (under `info/`).

> `conda index` iterates over the published `conda` packages, extracts the metadata and consolidates all the found JSON blobs into a single JSON file: `repodata.json`.

Reviewer 2: ```suggestion
`conda index` iterates over the published `conda` packages, extracts the metadata and consolidates all the found JSON blobs into a single JSON file: `repodata.json`.
This is where the hashes and file sizes are added too.
```

This is the metadata file that the CLI clients download initially to _solve_ the environment.

Since the metadata is external to the package files, some details can be modified without rebuilding packages, which simplifies some maintenance tasks notably.

Repodata patches are created in `conda-forge/conda-forge-repodata-patches-feedstock`, which generates (and uploads) a regular `conda` package as a result:
[`conda-forge-repodata-patches`](https://anaconda.org/conda-forge/conda-forge-repodata-patches/files).
Each of these timestamped packages contains the patch instructions for each channel subdir on conda-forge.

> The Anaconda infrastructure takes the JSON files from these packages and applies them on top of the vanilla `repodata.json`.

Reviewer 2: ```suggestion
The Anaconda infrastructure takes the JSON files from these packages and applies them on top of the vanilla `repodata.json` (also available for download as `repodata_from_packages.json`).
```

> Since this operates as a regular feedstock for package publication,

Reviewer 2: ```suggestion
Since `conda-forge-repodata-patches-feedstock` operates as a regular feedstock for package publication,
```

there are no further infrastructural details to cover.

#### Mark a package as broken

Sometimes a package is faulty in ways that a repodata patch cannot amend (e.g. bad binary).
In these cases, conda-forge does not remove packages from Anaconda.org.
Instead, it marks them with the `broken` label, which has a special meaning:

> packages labeled as such will be removed from the repodata via automated patches.

Reviewer 2: ```suggestion
packages labeled as such will be removed from the repodata via automated patches.
This action is reversible and doesn't change the direct URL of the artifact, which
can always be downloaded from e.g. a lockfile.
```
